File: ts/src/financial/risk.ts
/**
 * Risk metrics for financial analysis - WASM-backed implementations
 */

import { getWasmModule, ensureWasm, isWasmReady } from '../wasm-loader';

/**
 * Calculate daily returns from prices
 */
export function dailyReturns(prices: number[]): number[] {
  if (prices.length < 2) return [];
  const returns: number[] = [];
  for (let i = 1; i < prices.length; i++) {
    returns.push((prices[i] - prices[i - 1]) / prices[i - 1]);
  }
  return returns;
}

/**
 * Calculate cumulative returns from prices
 */
export function cumulativeReturns(prices: number[]): number[] {
  if (prices.length === 0) return [];
  const initial = prices[0];
  return prices.map((p) => (p - initial) / initial);
}

/**
 * Calculate annualized return
 * @param returns - Daily returns
 * @param tradingDays - Trading days per year (default: 252)
 */
export function annualizedReturn(returns: number[], tradingDays: number = 252): number {
  if (returns.length === 0) return 0;
  const meanDaily = returns.reduce((a, b) => a + b, 0) / returns.length;
  return meanDaily * tradingDays;
}

/**
 * Calculate annualized volatility
 * @param returns - Daily returns
 * @param tradingDays - Trading days per year (default: 252)
 */
export function annualizedVolatility(returns: number[], tradingDays: number = 252): number {
  if (returns.length < 2) return 0;
  const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
  const variance = returns.reduce((sum, r) => sum + (r - mean) ** 2, 0) / returns.length;
  return Math.sqrt(variance * tradingDays);
}

/**
 * Calculate Sharpe ratio (WASM-backed)
 * @param returns - Array of returns
 * @param riskFreeRate - Risk-free rate (default: 0)
 */
export async function sharpeRatio(
  returns: number[],
  riskFreeRate: number = 0
): Promise<number> {
  await ensureWasm();
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wasm = getWasmModule() as any;
  return wasm.compute_sharpe_ratio(new Float64Array(returns), riskFreeRate);
}

/**
 * Calculate Sharpe ratio synchronously (requires WASM to be initialized)
 * @param returns - Array of returns
 * @param riskFreeRate - Risk-free rate (default: 0)
 */
export function sharpeRatioSync(
  returns: number[],
  riskFreeRate: number = 0
): number {
  if (!isWasmReady()) {
    throw new Error('WASM not initialized. Call initWasm() first or use async sharpeRatio()');
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wasm = getWasmModule() as any;
  return wasm.compute_sharpe_ratio(new Float64Array(returns), riskFreeRate);
}

/**
 * Calculate Sortino ratio (WASM-backed)
 * @param returns - Array of returns
 * @param riskFreeRate - Risk-free rate (default: 0)
 */
export async function sortinoRatio(
  returns: number[],
  riskFreeRate: number = 0
): Promise<number> {
  await ensureWasm();
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wasm = getWasmModule() as any;
  return wasm.compute_sortino_ratio(new Float64Array(returns), riskFreeRate);
}

/**
 * Calculate Sortino ratio synchronously (requires WASM to be initialized)
 */
export function sortinoRatioSync(
  returns: number[],
  riskFreeRate: number = 0
): number {
  if (!isWasmReady()) {
    throw new Error('WASM not initialized. Call initWasm() first or use async sortinoRatio()');
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wasm = getWasmModule() as any;
  return wasm.compute_sortino_ratio(new Float64Array(returns), riskFreeRate);
}

/**
 * Calculate maximum drawdown (WASM-backed)
 * @param prices - Price series or equity curve
 */
export async function maxDrawdown(prices: number[]): Promise<number> {
  await ensureWasm();
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wasm = getWasmModule() as any;
  return wasm.compute_max_drawdown(new Float64Array(prices));
}

/**
 * Calculate maximum drawdown synchronously (requires WASM to be initialized)
 */
export function maxDrawdownSync(prices: number[]): number {
  if (!isWasmReady()) {
    throw new Error('WASM not initialized. Call initWasm() first or use async maxDrawdown()');
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wasm = getWasmModule() as any;
  return wasm.compute_max_drawdown(new Float64Array(prices));
}

/**
 * Calculate drawdown series
 * @param prices - Price series or equity curve
 */
export function drawdownSeries(prices: number[]): number[] {
  if (prices.length === 0) return [];

  const drawdowns: number[] = [];
  let peak = prices[0];

  for (const price of prices) {
    if (price > peak) {
      peak = price;
    }
    drawdowns.push((peak - price) / peak);
  }

  return drawdowns;
}

/**
 * Calculate Value at Risk using historical method (WASM-backed)
 * @param returns - Array of returns
 * @param confidence - Confidence level (e.g., 0.95 for 95%)
 */
export async function varHistorical(
  returns: number[],
  confidence: number = 0.95
): Promise<number> {
  await ensureWasm();
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wasm = getWasmModule() as any;
  return Math.abs(wasm.compute_var(new Float64Array(returns), confidence));
}

/**
 * Calculate Value at Risk synchronously (requires WASM to be initialized)
 */
export function varHistoricalSync(
  returns: number[],
  confidence: number = 0.95
): number {
  if (!isWasmReady()) {
    throw new Error('WASM not initialized. Call initWasm() first or use async varHistorical()');
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wasm = getWasmModule() as any;
  return Math.abs(wasm.compute_var(new Float64Array(returns), confidence));
}

/**
 * Calculate Conditional VaR (Expected Shortfall)
 * Note: Pure TypeScript implementation
 * @param returns - Array of returns
 * @param confidence - Confidence level (e.g., 0.95 for 95%)
 */
export function cvar(returns: number[], confidence: number = 0.95): number {
  if (returns.length === 0) return 0;

  const sorted = [...returns].sort((a, b) => a - b);
  const cutoffIndex = Math.floor((1 - confidence) * sorted.length);
  const tail = sorted.slice(0, cutoffIndex + 1);

  if (tail.length === 0) return 0;
  const avgLoss = tail.reduce((a, b) => a + b, 0) / tail.length;
  return -avgLoss; // Return as positive number
}

File: crates/rustful-financial/src/risk.rs
//! Risk metrics

/// Calculate Value at Risk (historical method)
pub fn var_historical(returns: &[f64], confidence: f64) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }
    let mut sorted = returns.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let index = ((1.0 - confidence) * sorted.len() as f64).floor() as usize;
    sorted.get(index).copied().unwrap_or(0.0)
}

/// Calculate Sharpe ratio
pub fn sharpe_ratio(returns: &[f64], risk_free_rate: f64) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }
    let mean: f64 = returns.iter().sum::<f64>() / returns.len() as f64;
    let excess = mean - risk_free_rate;
    let variance: f64 = returns.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / returns.len() as f64;
    let std_dev = variance.sqrt();
    if std_dev == 0.0 {
        0.0
    } else {
        excess / std_dev
    }
}

/// Calculate maximum drawdown
pub fn max_drawdown(equity_curve: &[f64]) -> f64 {
    if equity_curve.is_empty() {
        return 0.0;
    }
    let mut max_dd = 0.0;
    let mut peak = equity_curve[0];
    for &value in equity_curve {
        if value > peak {
            peak = value;
        }
        let dd = (peak - value) / peak;
        if dd > max_dd {
            max_dd = dd;
        }
    }
    max_dd
}

/// Calculate Sortino ratio (downside deviation only)
pub fn sortino_ratio(returns: &[f64], risk_free_rate: f64) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }
    let mean: f64 = returns.iter().sum::<f64>() / returns.len() as f64;
    let excess = mean - risk_free_rate;
    let downside: Vec<f64> = returns.iter().filter(|&&r| r < 0.0).copied().collect();
    if downside.is_empty() {
        return f64::INFINITY;
    }
    let downside_variance: f64 = downside.iter().map(|r| r.powi(2)).sum::<f64>() / downside.len() as f64;
    let downside_dev = downside_variance.sqrt();
    if downside_dev == 0.0 {
        f64::INFINITY
    } else {
        excess / downside_dev
    }
}
